### PR TITLE
Fix sparse setindex issue

### LINKED ISSUE #20657: Sparse matrix allows for duplicate entries
The representation of sparse matrices allow for duplicates. This seems prone to errors, for instance computing the number of non-zeros entries would be incorrect. 

Since this is not allowed in full matrices, does it make any sense to have this be the behavior for sparse matrices?

```
julia> a = spzeros(2, 3)
2×3 SparseMatrixCSC{Float64,Int64} with 0 stored entries

julia> a[[1,1,1,1,1,1],2]=1
1

julia> a
2×3 SparseMatrixCSC{Float64,Int64} with 6 stored entries:
  [1, 2]  =  1.0
  [1, 2]  =  1.0
  [1, 2]  =  1.0
  [1, 2]  =  1.0
  [1, 2]  =  1.0
  [1, 2]  =  1.0
```

### PR DESCRIPTION
**Pull request by fredrikekre**
~This can be done better after #20619 (or #20576 however that pans out) but at least this~ fix #20657

### CODE REVIEW

File: base/sparse/sparsematrix.jl [hunk 1]
```diff
@@ -2252,6 +2252,8 @@ function setindex!{Tv}(A::SparseMatrixCSC{Tv}, x::Number,
     if isempty(I) || isempty(J); return A; end
     if !issorted(I); I = sort(I); end
     if !issorted(J); J = sort(J); end
+    if !allunique(I); I = unique(I); end
```
Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: I'd put this before the `!issorted` checks and use `sort!(I)` instead of `I = sort(I)`

Comment (fredrikekre) [inline base/sparse/sparsematrix.jl]: But we can only use `sort!` in the case where there are duplicates then (such that we have `I = unique(I)` making a copy and then sort in place). The input vectors should not be mutated so `unique` or `sort` need to make a copy. I would think it is more common with unsorted vectors than vectors with duplicates, right? In that case, wouldn't this be better:

```julia
if !issorted(I)
    if !allunique(I)
        I = sort!(unique(I))
    else # unique but not sorted
        I = sort(I)
    end
elseif !allunique(I) # sorted
    I = unique(I)
end
```

Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: I wasn't seeing the whole function, but you're right, here `I` or `J` shouldn't be mutated. Another option that comes to mind then is

```julia
if !issorted(I, lt=≤); I = !allunique(I) ? sort!(unique(I)) : sort(I); end
```

Comment (fredrikekre) [inline base/sparse/sparsematrix.jl]: That's a nice solution. Thanks!

File: base/sparse/sparsematrix.jl [hunk 2]
```diff
@@ -2250,10 +2250,8 @@ setindex!(A::SparseMatrixCSC, x, i::Union{Integer, AbstractVector}, ::Colon) = s
 function setindex!{Tv}(A::SparseMatrixCSC{Tv}, x::Number,
         I::AbstractVector{<:Integer}, J::AbstractVector{<:Integer})
     if isempty(I) || isempty(J); return A; end
-    if !issorted(I); I = sort(I); end
-    if !issorted(J); J = sort(J); end
-    if !allunique(I); I = unique(I); end
-    if !allunique(J); J = unique(J); end
+    if !issorted(I, lt=≤); I = !allunique(I) ? sort!(unique(I)) : sort(I); end
```
Comment (tkelman) [inline base/sparse/sparsematrix.jl]: I think there was an issue on this a while back - probably a good idea to at least add a comment that `by=≤` is testing for strict sorting

**Review by Sacha0 — APPROVED** [review]

File: base/sparse/sparsematrix.jl (hunk 2 above)
Comment (Sacha0) [inline base/sparse/sparsematrix.jl]: #19280? Also ref #18281. Best!

File: base/sparse/sparsematrix.jl [hunk 3]
```diff
@@ -2250,8 +2250,9 @@ setindex!(A::SparseMatrixCSC, x, i::Union{Integer, AbstractVector}, ::Colon) = s
 function setindex!{Tv}(A::SparseMatrixCSC{Tv}, x::Number,
         I::AbstractVector{<:Integer}, J::AbstractVector{<:Integer})
     if isempty(I) || isempty(J); return A; end
-    if !issorted(I); I = sort(I); end
-    if !issorted(J); J = sort(J); end
+    # lt=≤ to check for strict sorting
+    if !issorted(I, lt=≤); I = !allunique(I) ? sort!(unique(I)) : sort(I); end
```
Comment (pabloferz) [inline base/sparse/sparsematrix.jl]: Thinking about this a bit more. The check for uniqueness is already allocating a `Set` and in any case you allocate a new array, so simply

```julia
if !issorted(I, lt=≤); I = sort!(unique(I)); end
```

should be good enough in general.

**Comment by fredrikekre** [comment]
Hmm, messed up the rebase and got the old version back. Should be correct now

**Review by Sacha0 — APPROVED** [review]
LGTM for the revision!

**Comment by StefanKarpinski** [comment]
It's rather unfortunate that `lt=≤` is how you express strictly monotonic ordering. It would be much more intuitive to have it indicated with `le=<` or something like that.

**Comment by Sacha0** [comment]
> It's rather unfortunate that lt=≤ is how you express strictly monotonic ordering. It would be much more intuitive to have it indicated with le=< or something like that.

Ref. #19280 and #18281. Best!

**Pull request merged**